### PR TITLE
Uefi boot support  on mkcloud

### PR DESCRIPTION
**Pull request by mmnelemane**
Allow support for firmware_type as a parameter to admin_config and add an option to mkcloud "want_efi" to use UEFI image to boot mkcloud.

### CODE REVIEW

File: scripts/mkcloud [hunk 1]
```diff
@@ -482,6 +491,13 @@ function setuphost()
         [[ $arch = s390x ]] && {
             echo 1 > /proc/sys/vm/allocate_pgste
         }
+
+        # enable UEFI boot
+        [[ -n $want_efi ]] && {
```
Comment (dirkmueller) [inline scripts/mkcloud]: either we rename that to want_x86_64_efi or we need to add another $arch condition here..

File: scripts/mkcloud [hunk 2]
```diff
@@ -482,6 +491,13 @@ function setuphost()
...
+            zypper --non-interactive in --no-recommends \
+                ovmf ovmf-tools qemu-ovmf-x86_64
```
Comment (dirkmueller) [inline scripts/mkcloud]: I think just setting 

```bash
    osloader=qemu-ovmf-x86_64 
```

should be enough - see previous aarch64 case.

File: scripts/mkcloud [hunk 3]
```diff
@@ -482,6 +491,13 @@ function setuphost()
...
+            [ "$?" == 0 -o "$?" == 4 ] || complain 10 "setuphost failed to install required packages"
```
Comment (dirkmueller) [inline scripts/mkcloud]: which should be fine, older host systems don't have ovmf available. I wouldn't make that a failure, but that can be debated.

File: scripts/qa_crowbarsetup.sh [hunk 4]
```diff
@@ -675,7 +675,7 @@ function addcloud6pool()
 function addcctdepsrepo()
 {
     if [[ "$cloudsource" == @(develcloud5|GM5|GM5+up) ]]; then
-        zypper ar -f http://$susedownload/ibs/Devel:/Cloud:/Shared:/Rubygem/SLE_11_SP3/Devel:Cloud:Shared:Rubygem.repo
+        zypper --non-interactive --no-gpg-checks ar -f http://$susedownload/ibs/Devel:/Cloud:/Shared:/Rubygem/SLE_11_SP3/Devel:Cloud:Shared:Rubygem.repo
```
Comment (dirkmueller) [inline scripts/qa_crowbarsetup.sh]: whats the point of --no-gpg-checks here?

File: scripts/lib/mkcloud-libvirt.sh [hunk 5]
```diff
@@ -103,6 +103,11 @@ function libvirt_prepare()
 
 function libvirt_setupadmin()
 {
-    ${mkcloud_lib_dir}/libvirt/admin-config $cloud $admin_node_memory $adminvcpus $emulator $admin_node_disk "$localreposdir_src" "$localreposdir_target" > /tmp/$cloud-admin.xml
+    nvramtemplate=""
```
Comment (dirkmueller) [inline scripts/lib/mkcloud-libvirt.sh]: why do you need to explicitely pass down the nvram template? that sounds wrong to me, libvirt has autodetection code for this, the default should be fine.

Comment (mmnelemane) [inline scripts/lib/mkcloud-libvirt.sh]: I understood that libvirt uses the NVRAM master from qemu.conf which is default and the "template" is used to override the master with the template needed for ovmf. I could change to not pass any template parameter for the general case and use the template file below for OVMF. Would that be better ?

File: scripts/mkcloud (hunk 1 above)
Comment (mmnelemane) [inline scripts/mkcloud]: Thanks. Will change it to use $arch in the next update to the PR.

File: scripts/qa_crowbarsetup.sh (hunk 4 above)
Comment (mmnelemane) [inline scripts/qa_crowbarsetup.sh]: It was an attempt to fix a local test failure which dint work anyways. Will remove in the next update to the PR. Thanks

File: scripts/lib/mkcloud-libvirt.sh (hunk 5 above)
Comment (dirkmueller) [inline scripts/lib/mkcloud-libvirt.sh]: Yes, because the template might be distribution specific and by not hardcoding we just use the autodetected default which should "work" always.

File: scripts/mkcloud [hunk 6]
```diff
@@ -483,6 +491,13 @@ function setuphost()
         [[ $arch = s390x ]] && {
             echo 1 > /proc/sys/vm/allocate_pgste
         }
+
+        # enable UEFI boot
+        [[ -n $want_efi ]] && [[ $arch = x86_64 ]] && {
+            zypper --non-interactive in --no-recommends \
+                ovmf ovmf-tools qemu-ovmf-x86_64
+            [ "$?" == 0 -o "$?" == 4 ] || complain 10 "setuphost failed to install required packages"
```
Comment (bmwiedemann) [inline scripts/mkcloud]: should use the ensure_packages_installed function to reduce code duplication

**Comment by bmwiedemann** [comment]
IMHO the python style cleanups should be split into a separate commit. Can be done with git reset --patch HEAD^ and git commit

File: scripts/lib/libvirt/fixtures/cloud-admin.xml [hunk 7]
```diff
@@ -5,7 +5,7 @@
   <vcpu>1</vcpu>
   <os>
     <type arch='x86_64' machine='pc-0.14'>hvm</type>
-
+
```
Comment (aplanas) [inline scripts/lib/libvirt/fixtures/cloud-admin.xml]: Extra spaces here

**Comment by aplanas** [comment]
The python part LGTM, but I do not tested it

+1

File: scripts/lib/libvirt/libvirt_setup.py [hunk 8]
```diff
@@ -56,15 +56,24 @@ def get_machine_arch():
...
+def get_os_loader(firmware_type=""):
     if 'aarch64' in get_machine_arch():
-        return """
-  <loader readonly='yes' type='pflash'>%s</loader>
-""" % path
+        path = '/usr/share/qemu/aavmf-aarch64-code.bin'
+        return """<loader readonly='yes' type='pflash'>%s</loader>""" % path
+    elif 'x86_64' in get_machine_arch() and firmware_type == "uefi":
+        path = '/usr/share/qemu/ovmf-x86_64-ms-code.bin'
+        return """<loader readonly='yes' type='pflash'>%s</loader>""" % path
```
Comment (bmwiedemann) [inline scripts/lib/libvirt/libvirt_setup.py]: Code duplication with line 62. Could be avoided by moving the return outside the if,
with something like
```python
if path=="":
   return ""
```
 coming first

Comment (aplanas) [inline scripts/lib/libvirt/libvirt_setup.py]: Actually I would avoid returning in multiple places, but is an small code. But I guess that @bmwiedemann suggestion is something like:

```
path = None
template = """<loader readonly='yes' type='pflash'>%s</loader>"""
if 'aarch64' in get_machine_arch():
    path = '/usr/share/qemu/aavmf-aarch64-code.bin'
elif x86_64' in get_machine_arch() and firmware_type == "uefi":
    path = '/usr/share/qemu/ovmf-x86_64-ms-code.bin'

return template%path if path else None
```

File: scripts/lib/libvirt/libvirt_setup.py [hunk 9]
```diff
@@ -56,13 +56,24 @@ def get_machine_arch():
...
     if 'aarch64' in get_machine_arch():
-        return """
-  <loader readonly='yes' type='pflash'>%s</loader>
-""" % path
-    return ""
+        path = "/usr/share/qemu/aavmf-aarch64-code.bin"
+        osloader = "<loader readonly='yes' type='pflash'>%s</loader>" % path
+    elif 'x86_64' in get_machine_arch() and firmware_type == "uefi":
+        path = "/usr/share/qemu/ovmf-x86_64-ms-code.bin"
+        osloader = "<loader readonly='yes' type='pflash'>%s</loader>" % path
```
Comment (bmwiedemann) [inline scripts/lib/libvirt/libvirt_setup.py]: https://github.com/SUSE-Cloud/automation/pull/1141#discussion_r75468641 is IMHO a nicer way to do this.

File: scripts/lib/libvirt/fixtures/cloud-admin.xml (hunk 7 above)
Comment (mmnelemane) [inline scripts/lib/libvirt/fixtures/cloud-admin.xml]: This was intentionally done to pass the test. Unfortunately, the tests do a byte-by-byte comparison to test the xml files generated. So the spaces ensure the tests dont throw exception due to space-mismatch. A better fix is to update the tests to compare xml using correct library which can be another PR IMO.

File: scripts/lib/libvirt/libvirt_setup.py (hunk 9 above)
Comment (mmnelemane) [inline scripts/lib/libvirt/libvirt_setup.py]: @bweidemann: Thanks. Have updated as mentioned by @aplanas.

**Comment by mmnelemane** [comment]
@aplanas, @bmwiedemann : The jenkins passes with and without the flag now. Also updated the comments on the Python part.

**Comment by nicolasbock** [comment]
+1

**Comment by aplanas** [comment]
Still +1

File: scripts/mkcloud [hunk 10]
```diff
@@ -428,10 +431,15 @@ function onhost_deploy_image()
     local role=$1
     local image=$(dist_to_image_name $2)
     local disk=$3
+    local image_url="http://$clouddata/images/$arch/$image"
+
+    [[ -n $want_efi ]] && {
+        image_url="http://download.suse.de/ibs/Devel:/Cloud:/Images/images_SLE_12_SP2/mkcloud-admin-node-SLE12SP2-uefi.x86_64.qcow2"
+        image="mkcloud-admin-node-SLE12SP2-uefi.x86_64.qcow2"
```
Comment (bmwiedemann) [inline scripts/mkcloud]: should be made more generic so that it can work on Cloud6 and Cloud8

Comment (mmnelemane) [inline scripts/mkcloud]: @bmwiedemann: The SP1 image cannot be built currently due to some unsupported dependencies (libsystemd, rubygem-cfa). So currently only SP2 image is possible.

**Comment by mjura** [comment]
LGTM

**Comment by mmnelemane** [comment]
@mjura: Thanks for pointing out on the firmware. It is using suse-* firmware now.

**Comment by mjura** [comment]
Thank you, still +1

File: scripts/lib/libvirt/templates/admin-node.xml [hunk 11]
```diff
@@ -5,8 +5,8 @@
   <vcpu>$adminvcpus</vcpu>
   <os>
     <type arch='$march' machine='$machine'>hvm</type>
-$osloader
-    <nvram>/var/lib/libvirt/$cloud-admin_VARS.fd</nvram>
+    $osloader
```
Comment (dirkmueller) [inline scripts/lib/libvirt/templates/admin-node.xml]: this was intentionally not properly indented in order to not produce trailing whitespace in case $osloader expands to empty.

File: scripts/lib/mkcloud-libvirt.sh [hunk 12]
```diff
@@ -103,6 +103,12 @@ function libvirt_prepare()
...
+    if [[ $(uname -m) == x86_64 ]] && [[ "$firmware_type" == "uefi" ]]; then
+        nvramtemplate="/usr/share/qemu/ovmf-x86_64-ms-vars.bin"
+        cp $nvramtemplate /var/lib/libvirt/qemu/nvram/cloud-admin_VARS.fd
```
Comment (dirkmueller) [inline scripts/lib/mkcloud-libvirt.sh]: thats wrong, libvirt takes care of that, the whole block can be deleted imho

File: scripts/mkcloud [hunk 13]
```diff
@@ -135,6 +136,7 @@ scp="scp $sshopts"
 : ${cct_tests:="features:base"}
 
 [[ -z ${want_mtu_size} ]] && want_mtu_size=1500
+iscloudver 6plus && [[ -n $want_efi ]] && firmware_type="uefi"
```
Comment (dirkmueller) [inline scripts/mkcloud]: probably this should check for "-a [[ $arch = x86_64 ]]"

File: scripts/mkcloud [hunk 14]
```diff
@@ -428,11 +431,16 @@ function onhost_deploy_image()
     local role=$1
     local image=$(dist_to_image_name $2)
     local disk=$3
+    local image_url="http://$clouddata/images/$arch/$image"
+
+    iscloudver 6plus && [[ -n $want_efi ]] && {
+        image_url="http://download.suse.de/ibs/Devel:/Cloud:/Images/images_SLE_12_SP2/mkcloud-admin-node-SLE12SP2-uefi.x86_64.qcow2"
```
Comment (dirkmueller) [inline scripts/mkcloud]: needs to be put on clouddata directly via a change to syncSLErepos script

File: scripts/mkcloud [hunk 15]
```diff
@@ -475,15 +483,22 @@ function setuphost()
...
-            dnsmasq netcat-openbsd ebtables libvirt-python
-        [ "$?" == 0 -o "$?" == 4 ] || complain 10 "setuphost failed to install required packages"
+
+        ensure_packages_installed libvirt $kvmpkg $osloader $ipxe \
+            lvm2 curl wget bridge-utils dnsmasq netcat-openbsd \
+            ebtables libvirt-python
+
+        # enable UEFI boot
+        [[ $arch = x86_64 ]] && [[ -n $want_efi ]] && {
+            zypper --non-interactive ar http://download.suse.de/ibs/Devel:/Virt:/SLE-12-SP2/SUSE_SLE-12-SP2_GA_standard/Devel:Virt:SLE-12-SP2.repo
```
Comment (dirkmueller) [inline scripts/mkcloud]: that would be wrong on SP1 hosts (e.g. cloud 6.x). I think easiest is to make that a cloud7 only feature and then we can remove this hack

File: scripts/mkcloud [hunk 16]
```diff
@@ -475,15 +483,22 @@ function setuphost()
...
+            ensure_packages_installed ovmf ovmf-tools qemu-ovmf-x86_64
+        }
 
         # enable KVM
         [[ $arch = s390x ]] && {
             echo 1 > /proc/sys/vm/allocate_pgste
         }
+
```
Comment (dirkmueller) [inline scripts/mkcloud]: extra white line?

File: scripts/mkcloud [hunk 17]
```diff
@@ -1338,6 +1354,8 @@ Optional
         Set to 1 to deploy sahara.
     want_postgresql=0 (default=1)
         Use PostgreSQL instead of SQLite as a Crowbar database
+    want_efi=1 (default=0)
+        Deploy nodes with EFI. Only available from SLE12_SP1 and later.
```
Comment (dirkmueller) [inline scripts/mkcloud]: maybe say "Cloud 7 and later" ?

File: scripts/lib/mkcloud-libvirt.sh (hunk 12 above)
Comment (mmnelemane) [inline scripts/lib/mkcloud-libvirt.sh]: @dirkmueller: libvirt can take care of this only if the `nvram` is explicitly specified in qemu.conf which is not done currently. If not, we need to provide the store as a template (as above). I get the below error if neither of this is done. AFAIU, the template can override the conf. Which do you think is a better option ?

```
    dom.create()
  File "/usr/lib64/python2.7/site-packages/libvirt.py", line 1035, in create
    if ret == -1: raise libvirtError ('virDomainCreate() failed', dom=self)
libvirt.libvirtError: operation failed: unable to find any master var store for loader: /usr/share/qemu/ovmf-x86_64-ms-code.bin
```

File: scripts/mkcloud (hunk 14 above)
Comment (mmnelemane) [inline scripts/mkcloud]: Added with https://github.com/SUSE-Cloud/automation/pull/1226

**Comment by mmnelemane** [comment]
Rebased and updated to use libvirt autodetect for nvram store. Also enabled efi only for Cloud 7 and later. So the bug in libvirt (SP1) for autodetect failure should not affect as Cloud7 is already moved to use SP2 (where the bug is fixed)

**Comment by dirkmueller** [comment]
@mmnelemane thanks, could you take a look at the unit test failure please?

File: scripts/lib/libvirt/libvirt_setup.py [hunk 18]
```diff
@@ -56,13 +56,14 @@ def get_machine_arch():
...
 
 
-def get_os_loader():
-    path = '/usr/share/qemu/aavmf-aarch64-code.bin'
+def get_os_loader(firmware_type=None):
+    path = None
+    template = "<loader readonly='yes' type='pflash'>%s</loader>"
     if 'aarch64' in get_machine_arch():
-        return """
-  <loader readonly='yes' type='pflash'>%s</loader>
```
Comment (dirkmueller) [inline scripts/lib/libvirt/libvirt_setup.py]: at least for the aarch64 case we actually still need the <nvram> element (and you also need it for the uefi case, otherwise reboot does not work)

File: scripts/lib/libvirt/templates/admin-node.xml [hunk 19]
```diff
@@ -5,8 +5,7 @@
...
```
Comment (dirkmueller) [inline scripts/lib/libvirt/templates/admin-node.xml]: we might not want to indent it because it would produce trailing whitespace when it expands to empty..

File: scripts/lib/libvirt/libvirt_setup.py (hunk 18 above)
Comment (mmnelemane) [inline scripts/lib/libvirt/libvirt_setup.py]: for both aarch64 and uefi with x86_64, we still return a valid element [here](https://github.com/SUSE-Cloud/automation/pull/1141/files/6402e6477e9abc5502b683475e4579d8192f040b#diff-c4e86c61b1673e20ce3773716936b40cR66) . Just the logic has changed due to an earlier comment.

File: scripts/lib/libvirt/templates/admin-node.xml (hunk 19 above)
Comment (mmnelemane) [inline scripts/lib/libvirt/templates/admin-node.xml]: sure. will revert the indent.

**Comment by mmnelemane** [comment]
@dirkmueller : The changes now includes both admin and compute nodes to run with EFI images. Please recheck. Thanks.

**Comment by bmwiedemann** [comment]
fails fetching /images/x86_64/SLE12-SP2-uefi.qcow2 because the name is  SLES12-SP2-uefi.qcow2 with an additional "S" after SLE

**Comment by mmnelemane** [comment]
@bmwiedemann: Thanks. fixed.

**Comment by aplanas** [comment]
Rebase?

File: scripts/lib/libvirt/test_libvirt_setup.py [hunk 20]
```diff
@@ -100,68 +100,79 @@ def default_test_args(args):
...
     def test_compute_config_with_raid(self):
-        args = Arguments()
-        default_test_args(args)
-        args.controller_raid_volumes = 2
-        args.cephvolumenumber = 2
-        cpu_flags = libvirt_setup.readfile(
-            "{0}/cpu-intel.xml".format(TEMPLATE_DIR))
-
-        should_config = libvirt_setup.readfile(
+        self.args.controller_raid_volumes = 2
```
Comment (dirkmueller) [inline scripts/lib/libvirt/test_libvirt_setup.py]: Why the change to self.args here? Shouldn't this  be 

```python
args = self.args
args.controller_raid_volumes = 2
```

**Comment by dirkmueller** [comment]
@mmnelemane rebase?

File: scripts/lib/libvirt/templates/admin-node.xml [hunk 21]
```diff
@@ -6,7 +6,7 @@
   <os>
     <type arch='$march' machine='$machine'>hvm</type>
 $osloader
-    <nvram>/var/lib/libvirt/$cloud-admin_VARS.fd</nvram>
+    <nvram>/var/lib/libvirt/cloud-admin_VARS.fd</nvram>
```
Comment (dirkmueller) [inline scripts/lib/libvirt/templates/admin-node.xml]: why the removal of $ here? I think $cloud is the variable in case of multiple clouds running in parallel on the same host.
